### PR TITLE
#21 Add basic ship health and reparation

### PR DESCRIPTION
**Pull request by Pater9**
Hi,

I tried to add basic health and shielding to the ship entity, feel free to reject this PR if it differs from your original thoughts. 

Any criticism and improvements are welcome and greatly appreciated.

Thank you 🥇

### CODE REVIEW

**Comment by ardliath** [comment]
This is amazing, you’ve saved me so much work - thank you!!!

I’ll go through the changes, I can see a few tweaks but generally it’s fantastic!

File: DeepSpace.Contracts/IShieldUpgrades.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿namespace DeepSpace.Contracts
+{
+    public interface IShieldUpgrades
+    {
+        string Name { get; set; }
+        int ShieldValue { get; set; }
+        float Weight { get; set; }
```
Comment (ardliath) [inline DeepSpace.Contracts/IShieldUpgrades.cs]: These should probably be doubles for consistency

File: DeepSpace.Contracts/IShieldUpgrades.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿namespace DeepSpace.Contracts
+{
+    public interface IShieldUpgrades
```
Comment (ardliath) [inline DeepSpace.Contracts/IShieldUpgrades.cs:3]: I really like the idea of shield upgrades - I hadn't thought about that!

File: DeepSpace/Controllers/ShipController.cs [hunk 3]
```diff
@@ -88,16 +89,34 @@ public async Task<MoveShipResponse> Move([FromBody] MoveShipRequest value)
...
+        }
 
-            //// DELETE api/values/5
-            //[HttpDelete("{id}")]
-            //public void Delete(int id)
-            //{
-            //}
+
+        [HttpPut]
+        [ActionName("Repair")]
```
Comment (ardliath) [inline DeepSpace/Controllers/ShipController.cs]: Can we make these posts with the signature /api/ship/repair/hull and /api/ship/repair/shields?

File: DeepSpace/Messages/RepairShipRequest.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace DeepSpace.Messages
+{
+    public class RepairShipRequest
+    {
+        [JsonProperty(PropertyName = "commandCode")]
+        public string CommandCode { get; set; }
+
+        [JsonProperty(PropertyName = "health")]
+        public double Health { get; set; }
```
Comment (ardliath) [inline DeepSpace/Messages/RepairShipRequest.cs]: As this is a UIless game players/developers will call this method directly. By giving them a health variable you're effectively asking them how much they'd like to heal. I'd much rather this heals by... let's say 10 points per call for the time being. That means a player needs to sit and repair for a little while before they can fight again

File: DeepSpace.Contracts/Ship.cs [hunk 5]
```diff
@@ -1,12 +1,39 @@
...
 
+        public double BaseHealth { get; set; } = 100; // Could be based on ship type for example.
+        private double Shield => ShieldUpgrades.Sum(su => su.ShieldValue);
+        private List<IShieldUpgrades> ShieldUpgrades { get; set; }
+
+        public double CurrentHealth { get; set; }
+
         public Location Location { get; set; }
         public Move Move { get; set; }
+        public void AddShieldUpgrade(IShieldUpgrades upgrade)
```
Comment (ardliath) [inline DeepSpace.Contracts/Ship.cs]: This class is saved directly into the database so it should just be a list of properties. Can you add these methods into ShipManager - that's where the BLL wants to live

**Review by ardliath — CHANGES_REQUESTED** [review]
Really appreciate your help, you've got some great ideas here! Thanks for your contribution

**Comment by Pater9** [comment]
Nice review, I'll work on it tomorrow if I have some spare time. Thanks! :)

**Review by ardliath — APPROVED** [review]

**Comment by ardliath** [comment]
This looks fantastic, I’m just waiting for the CI checks to run and I’ll merge in

Thanks for your help!

**Pull request merged**